ouroboros-consensus-cardano/src/unstable-cardano-tools/Cardano/Tools/Headers.hs

```diff
@@ -0,0 +1,89 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE TypeApplications #-}
+
+-- | Tooling to generate and validate (Praos) headers.
+module Cardano.Tools.Headers where
+
+import Cardano.Crypto.DSIGN (deriveVerKeyDSIGN)
+import Cardano.Crypto.VRF (VRFAlgorithm (deriveVerKeyVRF, hashVerKeyVRF))
+import Cardano.Ledger.Api (ConwayEra, StandardCrypto)
+import Cardano.Ledger.BaseTypes (BoundedRational (boundRational), PositiveUnitInterval, mkActiveSlotCoeff)
+import Cardano.Ledger.Coin (Coin (..))
+import Cardano.Ledger.Compactible (toCompact)
+import Cardano.Ledger.Keys (VKey (..), hashKey)
+import Cardano.Ledger.PoolDistr (IndividualPoolStake (..))
+import Cardano.Protocol.TPraos.OCert (ocertN)
+import Control.Monad.Except (runExcept)
+import qualified Data.Aeson as Json
+import qualified Data.ByteString.Lazy as LBS
+import Data.Char (isSpace)
+import qualified Data.Map as Map
+import Data.Maybe (fromJust)
+import qualified Data.Text as Text
+import Ouroboros.Consensus.Block (validateView)
+import Ouroboros.Consensus.Protocol.Praos (
+    Praos,
+    doValidateKESSignature,
+    doValidateVRFSignature,
+ )
+import Ouroboros.Consensus.Protocol.Praos.Header (Header, hbOCert, pattern Header)
+import Ouroboros.Consensus.Shelley.HFEras ()
+import Ouroboros.Consensus.Shelley.Ledger (
+    ShelleyBlock,
+    mkShelleyHeader,
+ )
+
+import Ouroboros.Consensus.Shelley.Protocol.Praos ()
+import Test.Ouroboros.Consensus.Protocol.Praos.Header (
+    GeneratorContext (..),
+    MutatedHeader (..),
+    Mutation (..),
+    Sample (..),
+    expectedError,
+    generateSamples,
+    header,
+    mutation,
+ )
+
+type ConwayBlock = ShelleyBlock (Praos StandardCrypto) (ConwayEra StandardCrypto)
+
+-- * Running Generator
+data Options = Options
+
+run :: Options -> IO ()
+run Options = do
+    sample <- generateSamples
+    LBS.putStr $ Json.encode sample <> "\n"
+
+data ValidationResult = Valid Mutation | Invalid Mutation String
+    deriving (Eq, Show)
+
+validate :: GeneratorContext -> MutatedHeader -> ValidationResult
+validate context MutatedHeader{header, mutation} =
+    case (runExcept $ validateKES >> validateVRF, mutation) of
+        (Left err, mut) | ctor err == expectedError mut -> Valid mut
```

Comment:
It seems preferable to avoid the stringification currently involved in this equivalence check.

Have you already assessed how easy it is to use the existing error types?

Reply:
Yes, this is already fixed in a newer version with `expectedError :: Mutation -> (PraosValidationErr -> Bool) -> Bool`